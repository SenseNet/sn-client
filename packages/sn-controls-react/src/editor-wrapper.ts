export { Editor as default } from '@sensenet/editor-react'
