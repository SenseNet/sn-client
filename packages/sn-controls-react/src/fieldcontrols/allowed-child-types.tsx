import {
  ClickAwayListener,
  createStyles,
  FormGroup,
  FormHelperText,
  IconButton,
  InputLabel,
  List,
  ListItem,
  ListItemIcon,
  ListItemSecondaryAction,
  ListItemText,
  makeStyles,
  Paper,
  TextField,
  Typography,
} from '@material-ui/core'
import { ConstantContent, ODataCollectionResponse } from '@sensenet/client-core'
import { deepMerge } from '@sensenet/client-utils'
import { ContentType, GenericContent } from '@sensenet/default-content-types'
import React, { useCallback, useEffect, useState } from 'react'
import { typeicons } from '../assets/icons'
import { ReactClientFieldSetting } from './client-field-setting'
import { renderIconDefault } from './icon'
import { defaultLocalization } from './localization'

const INPUT_PLACEHOLDER = 'Start typing to add another type'
const ITEM_HEIGHT = 48

const useStyles = makeStyles(() => {
  return createStyles({
    inputContainer: {
      padding: '2px 4px',
      alignItems: 'center',
      boxShadow: 'none',
      position: 'relative',
    },
    input: {
      marginLeft: 8,
      flex: 1,
    },
    button: {
      padding: 10,
    },
    listContainer: {
      position: 'absolute',
      top: '40px',
      maxHeight: ITEM_HEIGHT * 2.5,
      overflow: 'auto',
      zIndex: 10,
    },
    list: {},
    listItem: {
      margin: 0,
    },
    remove: {
      marginLeft: '8px',
    },
    ddIsOpened: {
      display: 'block',
    },
    ddIsClosed: {
      display: 'none',
    },
  })
})

type AllowedChildTypesClassKey = Partial<ReturnType<typeof useStyles>>

const compare = (a: GenericContent, b: GenericContent) => {
  if (a.Name < b.Name) {
    return -1
  }
  if (a.Name > b.Name) {
    return 1
  }
  return 0
}

/**
 * Field control that represents an AllowedChildTypes field. Available values will be populated from the FieldSettings.
 */
export const AllowedChildTypes: React.FC<ReactClientFieldSetting & { classes?: AllowedChildTypesClassKey }> = (
  props,
) => {
  const classes = useStyles(props)
  const localization = deepMerge(defaultLocalization.allowedChildTypes, props.localization?.allowedChildTypes)

  const [isOpened, setIsOpened] = useState(false)
  const [inputValue, setInputValue] = useState('')
  const [selected, setSelected] = useState<ContentType[]>([])
  const [allowedTypesOnCTD, setAllowedTypesOnCTD] = useState<ContentType[]>([])
  const [allCTDs, setAllCTDs] = useState<ContentType[]>([])
  const [currentSelected, setCurrentSelected] = useState<ContentType>()

  const filteredList = allCTDs.filter((ctd) => ctd.DisplayName?.toLowerCase().includes(inputValue.toLowerCase()))

  const getTypes = useCallback(
    (typeResults: ContentType[], allowedChildTypesFromCTD: ODataCollectionResponse<ContentType>) => {
      if (props.actionName === 'new') {
        return (allowedChildTypesFromCTD.d.results.length && allowedChildTypesFromCTD.d.results) || []
      }
      return typeResults.length ? typeResults : allowedChildTypesFromCTD.d.results
    },
    [props.actionName],
  )

  useEffect(() => {
    ;(async () => {
      try {
        if (!props.repository) {
          throw new Error('You must pass a repository to this control')
        }
        if (!props.content) {
          return
        }

        const result = await props.repository.load<GenericContent>({
          idOrPath: props.content.Path,
          oDataOptions: {
            select: ['EffectiveAllowedChildTypes'],
            expand: ['EffectiveAllowedChildTypes'],
          },
        })

        const allowedChildTypesFromCTD = await props.repository.allowedChildTypes.getFromCTD(props.content.Path)

        const typeResults = result.d.EffectiveAllowedChildTypes as ContentType[]
        const types = getTypes(typeResults, allowedChildTypesFromCTD)

        setSelected(types)
        setAllowedTypesOnCTD(allowedChildTypesFromCTD.d.results)
      } catch (error) {
        console.error(error.message)
      }
    })()
  }, [props.actionName, props.content, props.repository, getTypes])

  useEffect(() => {
    ;(async () => {
      try {
        if (!props.repository) {
          throw new Error('You must pass a repository to this control')
        }

        const result = (await props.repository.executeAction({
          idOrPath: ConstantContent.PORTAL_ROOT.Id,
          name: 'GetAllContentTypes',
          method: 'GET',
          oDataOptions: {
            select: ['Name', 'DisplayName', 'Icon'],
          },
        })) as ODataCollectionResponse<ContentType>

        setAllCTDs(result.d.results.sort(compare))
      } catch (error) {
        console.error(error.message)
      }
    })()
  }, [props.repository])

  const handleRemove = (item: GenericContent) => {
    if (selected.length > 1) {
      const newSelected = selected.filter((i) => item.Name !== i.Name)
      setSelected(newSelected)
      props.fieldOnChange?.(
        props.settings.Name,
        newSelected.map((newItem) => newItem.Name),
      )
    } else {
      const valuesOnCTD = allowedTypesOnCTD.map((allowedType) => allowedType.Name)
      setSelected(allowedTypesOnCTD)
      props.fieldOnChange?.(props.settings.Name, valuesOnCTD)
    }
  }

  const handleInputChange = (e: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement | HTMLSelectElement>) => {
    const term = e.target.value

    setInputValue(term)

    if (term.length === 0) {
      setCurrentSelected(undefined)
    }
  }

  const handleClickAway = () => {
    setIsOpened(false)
  }

  const handleSelect = (item: ContentType) => {
    setCurrentSelected(item)
    setInputValue(item.DisplayName || '')
    setIsOpened(false)
  }

  const handleOnClick = () => {
    setIsOpened(true)
  }

  const handleAddClick = () => {
    const newSelection = currentSelected ? [...selected, currentSelected] : selected
    setSelected(newSelection)
    props.fieldOnChange?.(
      props.settings.Name,
      newSelection.map((item) => item.Name),
    )

    setCurrentSelected(undefined)
    setInputValue('')
  }

  const renderMenuItem = (item: ContentType, select: (item: ContentType) => void) => (
    <ListItem key={item.Id} value={item.Id} onClick={() => select(item)}>
      <ListItemIcon style={{ margin: 0 }}>
        {props.renderIcon
          ? props.renderIcon(item.Icon ? item.Icon.toLowerCase() : 'contenttype')
          : renderIconDefault(
              item.Icon && typeicons[item.Icon.toLowerCase()]
                ? typeicons[item.Icon.toLowerCase()]
                : typeicons.contenttype,
            )}
      </ListItemIcon>
      <ListItemText primary={item.DisplayName} />
    </ListItem>
  )

  switch (props.actionName) {
    case 'edit':
    case 'new':
      return (
        <ClickAwayListener onClickAway={handleClickAway}>
          <div>
            <InputLabel shrink htmlFor={props.settings.Name} required={props.settings.Compulsory}>
              {props.settings.DisplayName}
            </InputLabel>
            <List dense={true} className={classes.list}>
              {selected.map((item, index) => (
                <ListItem key={index} className={classes.listItem}>
                  <ListItemIcon style={{ margin: 0 }}>
                    {props.renderIcon
                      ? props.renderIcon(item.Icon ? item.Icon.toLowerCase() : 'contenttype')
                      : renderIconDefault(
                          item.Icon && typeicons[item.Icon.toLowerCase()]
                            ? typeicons[item.Icon.toLowerCase()]
                            : typeicons.contenttype,
                        )}
                  </ListItemIcon>
                  <ListItemText primary={item.DisplayName} />
                  <ListItemSecondaryAction className={classes.remove}>
                    <IconButton aria-label="Remove" onClick={() => handleRemove(item)}>
                      {props.renderIcon ? props.renderIcon('delete') : renderIconDefault('delete')}
                    </IconButton>
                  </ListItemSecondaryAction>
                </ListItem>
              ))}
            </List>
            <div style={{ position: 'relative' }}>
              <FormGroup row className={classes.inputContainer}>
                <TextField
                  id={props.settings.Name}
                  autoComplete="off"
                  type="search"
                  onClick={handleOnClick}
                  onChange={handleInputChange}
                  placeholder={INPUT_PLACEHOLDER}
                  fullWidth={true}
                  value={inputValue}
                  className={classes.input}
                />
                <IconButton
                  color="primary"
                  className={classes.button}
                  disabled={currentSelected && currentSelected.Name.length > 0 ? false : true}
                  onClick={handleAddClick}>
                  {props.renderIcon ? props.renderIcon('add_circle') : renderIconDefault('add_circle')}
                </IconButton>
              </FormGroup>
              <Paper className={`${classes.listContainer} ${isOpened ? classes.ddIsOpened : classes.ddIsClosed}`}>
                <List>
                  {filteredList.length > 0 ? (
                    filteredList.map((item: any) => renderMenuItem(item, handleSelect))
                  ) : (
                    <ListItem>No hits</ListItem>
                  )}
                </List>
              </Paper>
              {!props.hideDescription && <FormHelperText>{props.settings.Description}</FormHelperText>}
            </div>
          </div>
        </ClickAwayListener>
      )
    case 'browse':
    default:
      return (
        <>
          <Typography variant="caption" gutterBottom={true}>
            {props.settings.DisplayName}
          </Typography>
          {selected.length ? (
            <List dense={true}>
              {selected.map((item, index) => (
                <ListItem key={index}>
                  <ListItemIcon style={{ margin: 0 }}>
                    {props.renderIcon
                      ? props.renderIcon(item.Icon ? item.Icon.toLowerCase() : 'contenttype')
                      : renderIconDefault(
                          item.Icon && typeicons[item.Icon.toLowerCase()]
                            ? typeicons[item.Icon.toLowerCase()]
                            : typeicons.contenttype,
                        )}
                  </ListItemIcon>
                  <ListItemText primary={item.DisplayName} />
                </ListItem>
              ))}
            </List>
          ) : (
            <Typography variant="body1" gutterBottom={true}>
              {localization.noValue}
            </Typography>
          )}
        </>
      )
  }
}
