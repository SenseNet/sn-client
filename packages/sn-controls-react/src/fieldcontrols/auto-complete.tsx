import { Typography } from '@material-ui/core'
import { deepMerge } from '@sensenet/client-utils'
import { GenericContent, ReferenceFieldSetting } from '@sensenet/default-content-types'
import { Query } from '@sensenet/query'
import { ReferenceField } from '@sensenet/search-react'
import React from 'react'
import { changeTemplatedValue } from '../helpers'
import { defaultLocalization } from './localization'
import { ReactClientFieldSetting } from '.'

/**
 * Represents an autocomplete component
 */
export const AutoComplete: React.FC<ReactClientFieldSetting<ReferenceFieldSetting>> = (props) => {
  const localization = deepMerge(defaultLocalization.autoComplete, props.localization?.autoComplete)

  const defaultValue =
    (props.fieldValue && (props.fieldValue as any)[0].Id) || changeTemplatedValue(props.settings.DefaultValue)
  const fetchItems = async (fetchQuery: Query<GenericContent>) => {
    try {
      if (!props.repository) {
        throw new Error('You must pass a repository to this control')
      }

      const result = await props.repository.loadCollection<GenericContent>({
        path: '/Root',
        oDataOptions: {
          query: fetchQuery.toString(),
          select: 'all',
        },
      })
      return result.d.results
    } catch (error) {
      console.error(error.message)
    }
  }

  switch (props.actionName) {
    case 'edit':
    case 'new':
      return (
        <ReferenceField
          fieldSetting={props.settings}
          fieldName={props.settings.Name as any}
          defaultValueIdOrPath={defaultValue}
          onChange={(item) => props.fieldOnChange?.(props.settings.Name, item)}
          fetchItems={fetchItems as any}
          triggerClear={props.triggerClear}
          autoFocus={props.autoFocus}
        />
      )
    case 'browse':
    default: {
      return (
        <div>
          <Typography variant="caption" gutterBottom={true}>
            {props.settings.DisplayName}
          </Typography>
          <Typography variant="body1" gutterBottom={true}>
            {props.fieldValue ? (props.fieldValue as any)[0].DisplayName : localization.noValue}
          </Typography>
        </div>
      )
    }
  }
}
