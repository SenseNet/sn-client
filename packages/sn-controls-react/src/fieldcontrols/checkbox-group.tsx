/**
 * @module FieldControls
 */
import {
  Checkbox,
  FormControl,
  FormControlLabel,
  FormGroup,
  FormHelperText,
  FormLabel,
  TextField,
  Typography,
} from '@material-ui/core'
import { deepMerge } from '@sensenet/client-utils'
import { ChoiceFieldSetting } from '@sensenet/default-content-types'
import React, { useState } from 'react'
import { changeTemplatedValue } from '../helpers'
import { ReactClientFieldSetting } from './client-field-setting'
import { defaultLocalization } from './localization'

/**
 * Field control that represents a Choice field. Available values will be populated from the FieldSettings.
 */
export const CheckboxGroup: React.FC<ReactClientFieldSetting<ChoiceFieldSetting>> = (props) => {
  const localization = deepMerge(defaultLocalization.checkboxGroup, props.localization?.checkboxGroup)

  const getInitialstate = () => {
    if (!props.fieldValue) {
      if (props.actionName === 'new') {
        if (props.settings.DefaultValue) {
          const defaultValue = changeTemplatedValue(props.settings.DefaultValue)!.split(/,|;/)
          return props.settings.AllowMultiple ? defaultValue : defaultValue.slice(0, 1)
        }

        if (props.settings.Options?.length) {
          const selectedOnCtd = props.settings.Options.reduce<string[]>((selection, option) => {
            if (option.Selected) {
              selection.push(option.Value)
            }
            return selection
          }, [])
          const fieldValue = props.settings.AllowMultiple ? selectedOnCtd : selectedOnCtd.slice(0, 1)
          props.fieldOnChange?.(props.settings.Name, fieldValue)
          return fieldValue
        }
      }
      return ['']
    }
    if (!Array.isArray(props.fieldValue)) {
      return [props.fieldValue]
    }
    return props.fieldValue
  }

  const [value, setValue] = useState(getInitialstate)

  const handleChange = (name: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = props.settings.AllowMultiple
      ? event.target.checked
        ? [...value, name]
        : value.filter((item) => item !== name)
      : event.target.checked
      ? [name]
      : []

    setValue(newValue)
    props.fieldOnChange?.(props.settings.Name, newValue)
  }

  switch (props.actionName) {
    case 'new':
    case 'edit':
      return (
        <FormControl
          disabled={props.settings.ReadOnly}
          component={'fieldset' as 'div'}
          required={props.settings.Compulsory}>
          <FormLabel style={{ transform: 'translate(0, 1.5px) scale(0.75)', transformOrigin: 'top left' }}>
            {props.settings.DisplayName}
          </FormLabel>
          <FormGroup>
            {props.settings.Options?.map((option) => {
              return (
                <FormControlLabel
                  key={option.Value}
                  control={
                    <Checkbox
                      checked={value.includes(option.Value)}
                      onChange={handleChange(option.Value)}
                      value={option.Value}
                    />
                  }
                  label={option.Text}
                />
              )
            })}
          </FormGroup>
          {props.settings.AllowExtraValue ? <TextField placeholder="Extra value" /> : null}
          {!props.hideDescription && <FormHelperText>{props.settings.Description}</FormHelperText>}
        </FormControl>
      )
    case 'browse':
    default: {
      return (
        <>
          <Typography variant="caption" gutterBottom={true}>
            {props.settings.DisplayName}
          </Typography>
          <FormGroup>
            {props.fieldValue ? (
              Array.isArray(props.fieldValue) ? (
                props.fieldValue.length ? (
                  props.fieldValue.map((val: any, index: number) => (
                    <Typography variant="body1" gutterBottom={index === props.fieldValue!.length - 1} key={index}>
                      {props.settings.Options?.find((item) => item.Value === val)?.Text ?? ''}
                    </Typography>
                  ))
                ) : (
                  <Typography variant="body1" gutterBottom={true}>
                    {localization.noValue}
                  </Typography>
                )
              ) : (
                <Typography variant="body1" gutterBottom={true}>
                  {props.settings.Options?.find((item) => item.Value === (props.fieldValue as string))?.Text ?? ''}
                </Typography>
              )
            ) : (
              <Typography variant="body1" gutterBottom={true}>
                {localization.noValue}
              </Typography>
            )}
          </FormGroup>
        </>
      )
    }
  }
}
