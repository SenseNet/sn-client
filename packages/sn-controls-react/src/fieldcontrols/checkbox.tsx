/**
 * @module FieldControls
 */
import { FormControl, FormControlLabel, FormHelperText, Checkbox as MuiCheckbox } from '@material-ui/core'
import { deepMerge } from '@sensenet/client-utils'
import { FieldSetting } from '@sensenet/default-content-types'
import React, { useState } from 'react'
import { changeTemplatedValue } from '../helpers'
import { ReactClientFieldSetting } from './client-field-setting'
import { renderIconDefault } from './icon'
import { defaultLocalization } from './localization'

/**
 * Field control that represents a Checkbox field.
 */
export const Checkbox: React.FC<ReactClientFieldSetting<FieldSetting>> = (props) => {
  const localization = deepMerge(defaultLocalization.checkbox, props.localization?.checkbox)

  const initialState =
    props.fieldValue == null && props.actionName === 'new'
      ? changeTemplatedValue(props.settings.DefaultValue)?.toLowerCase() === 'true'
      : !!props.fieldValue
  const [value, setValue] = useState(initialState)

  const handleChange = () => {
    setValue(!value)
    props.fieldOnChange?.(props.settings.Name, !value)
  }

  switch (props.actionName) {
    case 'edit':
    case 'new':
      return (
        <FormControl required={props.settings.Compulsory} disabled={props.settings.ReadOnly}>
          <FormControlLabel
            name={props.settings.Name}
            control={<MuiCheckbox checked={value} onChange={handleChange} />}
            label={props.settings.DisplayName}
          />
          {!props.hideDescription && <FormHelperText>{props.settings.Description}</FormHelperText>}
        </FormControl>
      )
    case 'browse':
    default:
      return (
        <div style={{ display: 'flex' }}>
          <span style={{ marginRight: '0.5rem' }}>{props.settings.DisplayName}</span>
          {props.fieldValue != null
            ? props.renderIcon
              ? props.renderIcon(props.fieldValue ? 'check' : 'not_interested')
              : renderIconDefault(props.fieldValue ? 'check' : 'not_interested')
            : localization.noValue}
        </div>
      )
  }
}
