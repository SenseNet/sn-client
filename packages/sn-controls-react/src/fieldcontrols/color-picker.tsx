/**
 * @module FieldControls
 */
import { ClickAwayListener, FormControl, Icon, InputAdornment, TextField, Typography } from '@material-ui/core'
import { deepMerge } from '@sensenet/client-utils'
import { ColorFieldSetting } from '@sensenet/default-content-types'
import React, { useState } from 'react'
import { ColorResult, SketchPicker } from 'react-color'
import { changeTemplatedValue } from '../helpers'
import { ReactClientFieldSetting } from './client-field-setting'
import { defaultLocalization } from './localization'

const style = {
  input: {
    width: 80,
  },
  pickerContainer: {
    position: 'absolute',
    left: '24px',
    top: '20px',
    zIndex: 10,
  } as React.CSSProperties,
}

const renderIconDefault = (name: string, color: string) => {
  return <Icon style={{ color }}>{name}</Icon>
}

/**
 * Field control that represents a Color field. Available values will be populated from the FieldSettings.
 */
export const ColorPicker: React.FC<ReactClientFieldSetting<ColorFieldSetting>> = (props) => {
  const localization = deepMerge(defaultLocalization.colorPicker, props.localization?.colorPicker)
  const [value, setValue] = useState(
    props.fieldValue ?? (props.actionName === 'new' ? changeTemplatedValue(props.settings.DefaultValue) || '' : ''),
  )
  const [isPickerOpen, setIsPickerOpen] = useState(false)

  const handleChange = (color: ColorResult) => {
    props.fieldOnChange?.(props.settings.Name, color.hex)
    setValue(color.hex)
  }

  switch (props.actionName) {
    case 'edit':
    case 'new':
      return (
        <FormControl>
          <TextField
            label={props.settings.DisplayName}
            type="text"
            name={props.settings.Name}
            id={props.settings.Name}
            placeholder={props.settings.DisplayName}
            required={props.settings.Compulsory}
            disabled={props.settings.ReadOnly}
            value={value}
            onClick={() => setIsPickerOpen(true)}
            helperText={props.hideDescription ? undefined : props.settings.Description}
            InputProps={{
              startAdornment: (
                <InputAdornment position="start">
                  {props.renderIcon ? props.renderIcon('lens') : renderIconDefault('lens', value || 'transparent')}
                </InputAdornment>
              ),
            }}
          />
          {isPickerOpen ? (
            <ClickAwayListener onClickAway={() => setIsPickerOpen(false)}>
              <div style={style.pickerContainer}>
                <SketchPicker
                  color={value}
                  onChangeComplete={handleChange}
                  onSwatchHover={handleChange}
                  disableAlpha={true}
                />
              </div>
            </ClickAwayListener>
          ) : null}
        </FormControl>
      )
    case 'browse':
    default:
      return (
        <>
          <Typography variant="caption" gutterBottom={true}>
            {props.settings.DisplayName}
          </Typography>
          {props.fieldValue ? (
            <div style={{ display: 'flex' }}>
              {props.renderIcon ? props.renderIcon('lens') : renderIconDefault('lens', value)}
              <Typography variant="body1" gutterBottom={true} style={{ marginLeft: '0.5rem' }}>
                {value}
              </Typography>
            </div>
          ) : (
            <Typography variant="body1" gutterBottom={true} style={{ marginLeft: '0.5rem' }}>
              {localization.noValue}
            </Typography>
          )}
        </>
      )
  }
}
