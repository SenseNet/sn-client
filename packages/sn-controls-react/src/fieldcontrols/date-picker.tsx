/**
 * @module FieldControls
 */
import { createStyles, FormHelperText, makeStyles, Theme, Tooltip, Typography } from '@material-ui/core'
import { DateTimePicker, DatePicker as MUIDatePicker, MuiPickersUtilsProvider } from '@material-ui/pickers'
import type { MaterialUiPickersDate } from '@material-ui/pickers/typings/date'
import { deepMerge } from '@sensenet/client-utils'
import { DateTimeFieldSetting, DateTimeMode } from '@sensenet/default-content-types'
import DateFnsUtils from '@date-io/date-fns'
import intlFormatDistance from 'date-fns/intlFormatDistance'
import React, { useState } from 'react'
import { changeTemplatedValue } from '../helpers'
import { ReactClientFieldSetting } from './client-field-setting'
import { defaultLocalization } from './localization'
const minDatePickerDate = new Date('0001-01-01')

export const dateTimeOptions: Intl.DateTimeFormatOptions = {
  year: 'numeric',
  month: 'numeric',
  day: 'numeric',
  hour: '2-digit',
  minute: '2-digit',
  second: '2-digit',
}

const useStyles = makeStyles((theme: Theme) => {
  return createStyles({
    textDate: {
      color: `hsl(174deg 3% ${theme.palette.type === 'light' ? '41' : '74'}%)`,
      fontSize: '0.66rem',
      letterSpacing: '0.5px',
      marginLeft: '5px',
      verticalAlign: 'middle',
    },
  })
})

const initialValueState = ({
  fieldValue,
  actionName,
  settings,
}: Pick<ReactClientFieldSetting<DateTimeFieldSetting>, 'fieldValue' | 'actionName' | 'settings'>) => {
  if (fieldValue === '0001-01-01T00:00:00Z') {
    return null
  }

  if (fieldValue) {
    return fieldValue
  }

  if (actionName !== 'new') {
    return null
  }

  const secureCheckedDateInput = changeTemplatedValue(settings.DefaultValue, settings.EvaluatedDefaultValue)

  return secureCheckedDateInput
}
/**
 * Field control that represents a Date field. Available values will be populated from the FieldSettings.
 */
export const DatePicker: React.FC<ReactClientFieldSetting<DateTimeFieldSetting>> = (props) => {
  const classes = useStyles()

  const { settings, actionName, fieldValue, locale, localization, hideDescription, fieldOnChange } = props

  const localizationMerged = deepMerge(defaultLocalization.datePicker, localization?.datePicker)

  const [value, setValue] = useState(initialValueState({ fieldValue, actionName, settings }))

  const handleDateChange = (date: MaterialUiPickersDate) => {
    if (!date) {
      return
    }
    setValue(new Date(date).toISOString())
    fieldOnChange?.(settings.Name, new Date(date).toISOString())
  }

  const localeCode = locale?.code || window.navigator.language

  const dateFieldValue: Date = new Date(fieldValue as string)

  switch (actionName) {
    case 'edit':
    case 'new':
      return (
        <MuiPickersUtilsProvider utils={DateFnsUtils} locale={locale}>
          <>
            {settings.DateTimeMode === DateTimeMode.Date ? (
              <MUIDatePicker
                style={{ display: 'inherit' }}
                value={value}
                minDate={minDatePickerDate}
                onChange={handleDateChange}
                name={settings.Name}
                label={settings.DisplayName}
                id={settings.Name}
                disabled={settings.ReadOnly}
                InputLabelProps={{ shrink: true }}
                required={settings.Compulsory}
                format="yyyy MMMM dd"
              />
            ) : (
              <DateTimePicker
                style={{ display: 'inherit' }}
                minDate={minDatePickerDate}
                value={value}
                onChange={handleDateChange}
                label={settings.DisplayName}
                name={settings.Name}
                id={settings.Name}
                disabled={settings.ReadOnly}
                InputLabelProps={{ shrink: true }}
                required={settings.Compulsory}
                format="yyyy MMMM dd hh:mm aaaa"
              />
            )}
            {!hideDescription && <FormHelperText>{settings.Description}</FormHelperText>}
          </>
        </MuiPickersUtilsProvider>
      )
    default:
      return (
        <div>
          <Typography variant="caption" gutterBottom={true}>
            {settings.DisplayName}
            {settings.DateTimeMode === DateTimeMode.DateAndTime && fieldValue && (
              <span className={classes.textDate}>
                {intlFormatDistance(dateFieldValue, new Date(), {
                  locale: localeCode,
                })}
              </span>
            )}
          </Typography>

          <Tooltip title={fieldValue as string}>
            <Typography variant="body1" gutterBottom={true}>
              {fieldValue
                ? settings.DateTimeMode === DateTimeMode.Date
                  ? new Intl.DateTimeFormat(localeCode).format(dateFieldValue)
                  : new Intl.DateTimeFormat(localeCode, dateTimeOptions).format(dateFieldValue)
                : localizationMerged.noValue}
            </Typography>
          </Tooltip>
        </div>
      )
  }
}
