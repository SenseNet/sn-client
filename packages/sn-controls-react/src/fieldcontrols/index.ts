/**
 * @module FieldControls
 * @preferred
 * @description This module contains controls that are used to display or edit specific fields of a specified Content.
 */

export * from './number'
export * from './checkbox-group'
export * from './date-picker'
export * from './time-picker'
export * from './dropdown-list'
export * from './radio-button-group'
export * from './short-text'
export * from './tags-input'
export * from './textarea'
export * from './rich-text-editor'
export * from './name'
export * from './password'
export * from './file-name'
export * from './auto-complete'
export * from './file-upload'
export * from './avatar/avatar'
export * from './avatar/default-avatar-template'
export * from './reference-grid'
export * from './checkbox'
export * from './switcher'
export * from './empty-field-control'
export * from './color-picker'
export * from './allowed-child-types'
export * from './client-field-setting'
export * from './icon'
export * from './localization'
export * from './file-size'
export * from './page-count'
export * from './html-editor'
