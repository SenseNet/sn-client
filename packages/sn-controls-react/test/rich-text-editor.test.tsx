/* eslint-disable no-irregular-whitespace */

import { Editor } from '@sensenet/editor-react'
import FormHelperText from '@material-ui/core/FormHelperText'
import InputLabel from '@material-ui/core/InputLabel'
import Typography from '@material-ui/core/Typography'
import { mount, shallow } from 'enzyme'
import React, { FC, useEffect, useState } from 'react'
import { act } from 'react-dom/test-utils'
import { defaultLocalization, RichTextEditor } from '../src/fieldcontrols'

// The test must wait for React.lazy to resolve
const waitForComponentToPaint = async (wrapper) => {
  await act(async () => {
    await new Promise((resolve) => setTimeout(resolve))
    wrapper.update()
  })
}

beforeAll(() => {
  // React lazy and Suspense are not supported by server-side rendering, so it must be mocked
  // https://stackoverflow.com/a/65751177
  jest.mock('react', () => {
    const react = jest.requireActual('react')
    const Suspense = ({ children }) => {
      return children
    }

    const lazy = jest.fn().mockImplementation((fn) => {
      const Component: FC = (props) => {
        type DefaultExportedFC = { default: FC }
        const [C, setC] = useState<DefaultExportedFC>()

        useEffect(() => {
          fn().then((v: DefaultExportedFC) => {
            setC(v)
          })
        }, [])

        return C ? <C.default {...props} /> : null
      }

      return Component
    })

    return {
      ...react,
      lazy,
      Suspense,
    }
  })
})

describe('Rich text editor field control', () => {
  const defaultSettings = {
    Name: 'Description',
    Type: 'LongTextFieldSetting',
    DisplayName: 'Description of the field',
    FieldClassName: 'SenseNet.ContentRepository.Fields.LongTextField',
  }
  describe('in browse view', () => {
    it('should show the displayname and fieldValue when fieldValue is provided', () => {
      const value = '<h1>Hello World</h1>'
      const wrapper = shallow(<RichTextEditor fieldValue={value} actionName="browse" settings={defaultSettings} />)
      expect(wrapper.find(Typography).text()).toBe(defaultSettings.DisplayName)
      expect(wrapper.find('div').last().html()).toBe(`<div>${value}</div>`)
      expect(wrapper).toMatchSnapshot()
    })
    it('should show no value message when field value is not provided', () => {
      const wrapper = shallow(<RichTextEditor actionName="browse" settings={defaultSettings} />)
      expect(wrapper.find(Typography).last().text()).toBe(defaultLocalization.richTextEditor.noValue)
    })
  })
  describe('in edit/new view', () => {
    it('should set all the props', async () => {
      const value = 'Hello World'
      const wrapper = mount(
        <RichTextEditor
          fieldValue={value}
          actionName="edit"
          settings={{
            ...defaultSettings,
            ReadOnly: true,
            Compulsory: true,
            Description: 'description',
          }}
        />,
      )
      await waitForComponentToPaint(wrapper)

      expect(wrapper.find(Editor).prop('content')).toBe(value)
      expect(wrapper.find(Editor).prop('placeholder')).toBe(defaultSettings.DisplayName)
      expect(wrapper.find(Editor).prop('readOnly')).toBeTruthy()
      expect(wrapper.find(InputLabel).text()).toBe(`${defaultSettings.DisplayName}â€‰*`)
      expect(wrapper.find(InputLabel).prop('required')).toBeTruthy()
      expect(wrapper.find(FormHelperText).text()).toBe('description')
    })

    it('should set default value', async () => {
      const wrapper = mount(
        <RichTextEditor
          actionName="new"
          settings={{
            ...defaultSettings,
            DefaultValue: 'defaultValue',
          }}
        />,
      )
      await waitForComponentToPaint(wrapper)

      expect(wrapper.find(Editor).prop('content')).toBe('defaultValue')
    })
  })
})
