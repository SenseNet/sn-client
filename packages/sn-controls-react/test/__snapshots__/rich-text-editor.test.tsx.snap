// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rich text editor field control in browse view should show the displayname and fieldValue when fieldValue is provided 1`] = `
<div>
  <WithStyles(ForwardRef(Typography))
    gutterBottom={true}
    variant="caption"
  >
    Description of the field
  </WithStyles(ForwardRef(Typography))>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>Hello World</h1>",
      }
    }
  />
</div>
`;
